src/hotspot/share/gc/g1/g1ConcurrentRefineThreadsNeeded.cpp

```diff
@@ -0,0 +1,140 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/g1/g1Analytics.hpp"
+#include "gc/g1/g1ConcurrentRefineThreadsNeeded.hpp"
+#include "gc/g1/heapRegion.hpp"
+#include "gc/g1/g1Policy.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include <math.h>
+
+G1ConcurrentRefineThreadsNeeded::G1ConcurrentRefineThreadsNeeded(G1Policy* policy,
+                                                                 double update_period_ms) :
+  _policy(policy),
+  _update_period_ms(update_period_ms),
+  _predicted_time_until_next_gc_ms(0.0),
+  _predicted_cards_at_next_gc(0),
+  _threads_needed(0)
+{}
+
+// Estimate how many concurrent refinement threads we need to run to achieve
+// the target number of card by the time the next GC happens.  There are
+// several additional desirements we'd like to achieve while meeting that
+// goal.
+//
+// 1. Minimize the number of refinement threads running at once.
+//
+// 2. Minimize the number of activations and deactivations for the
+// refinement threads that run.
+//
+// 3. Delay performing refinement work.  Having more dirty cards waiting to
+// be refined can be beneficial, as further writes to the same card don't
+// create more work.
+void G1ConcurrentRefineThreadsNeeded::update(uint active_threads,
+                                             size_t available_bytes,
+                                             size_t num_cards,
+                                             size_t target_num_cards) {
+  const G1Analytics* analytics = _policy->analytics();
+
+  // Estimate time until next GC, based on remaining bytes available for
+  // allocation and the allocation rate.
+  double alloc_region_rate = analytics->predict_alloc_rate_ms();
+  double alloc_bytes_rate = alloc_region_rate * HeapRegion::GrainBytes;
+  if (alloc_bytes_rate == 0.0) {
+    _predicted_time_until_next_gc_ms = 0.0;
+  } else {
+    // If the heap size is large and the allocation rate is small, we can get
+    // a predicted time until next GC that is so large it can cause problems
+    // (such as overflow) in other calculations.  Limit the prediction to one
+    // hour, which is still large in this context.
+    const double one_hour_ms = 60.0 * 60.0 * MILLIUNITS;
+    double raw_time_ms = available_bytes / alloc_bytes_rate;
+    _predicted_time_until_next_gc_ms = MIN2(raw_time_ms, one_hour_ms);
+  }
+
+  // Estimate number of cards that need to be processed before next GC.  There
+  // are no incoming cards when time is short, because the controller activates
+  // refinement by mutator threads when there to a GC, to stay on target even
```

Comment:
"refinement by mutator threads when there to a GC" I do not understand this part of the sentence, what does "there to a gc" mean?

Maybe this should means:

"..., because in this case the controller activates refinement by mutator threads to stay on target even..."

Reply:
Done.  Something like that was intended.  Not sure what happened, maybe I got interrupted in the middle of editing.